File: web/src/core/metrics.ts
// Aggregates kinematics + drivetrain + dynamics into the live metrics the UI
// displays. Pure functions of (config, gearbox overrides, pose, payload) —
// this is the "digital twin evaluation" layer.

import type { GearboxParams, JointName, MotorParams, RobotConfig } from './config';
import { JOINT_NAMES } from './config';
import {
  availableJointTorque,
  backlashTcpError,
  jointResolution,
  maxJointSpeed,
  maxTcpSpeed,
  reflectedInertia,
} from './drivetrain';
import { gravityTorques, worstCaseLinkInertia } from './dynamics';
import type { JointAngles } from './kinematics';
import { jacobian, singularityMeasure } from './kinematics';
import type { TrajectoryPlan } from './trajectory';
import { sampleTrajectory } from './trajectory';

export interface JointMetrics {
  name: JointName;
  availableTorque: number;   // N·m at standstill
  requiredTorque: number;    // N·m gravity load at current pose
  utilization: number;       // required / available (∞-safe)
  maxSpeed: number;          // rad/s
  maxAccel: number;          // rad/s², torque budget / total inertia
  reflectedInertia: number;  // kg·m²
  linkInertia: number;       // kg·m² (worst-case pose)
  resolution: number;        // rad per microstep
  holdFails: boolean;        // cannot even hold this pose statically
}

export interface TwinMetrics {
  joints: JointMetrics[];
  backlashErrorTcp: number;  // m, worst case at current pose
  maxTcpSpeed: number;       // m/s bound at current pose
  singularity: number;       // 0 (singular) .. 1
  vmax: [number, number, number];
  amax: [number, number, number];
}

/** Fraction of the static torque budget reserved for acceleration. */
const ACCEL_TORQUE_FRACTION = 0.7;

export interface TrajectoryAudit {
  peakUtilization: number; // worst required/available torque ratio over the move
  peakJoint: JointName;
  skippedSteps: boolean;   // >1 → open-loop steppers would lose position
}

/**
 * Predictive torque audit of a planned move: at each sample, gravity plus
 * inertial torque is compared against what the drivetrain can deliver at that
 * joint speed. Deterministic — evaluated at plan time, not during playback.
 */
export function auditTrajectory(
  config: RobotConfig,
  gearboxes: Record<JointName, GearboxParams>,
  plan: TrajectoryPlan,
  payload: number,
  samples = 120,
): TrajectoryAudit {
  const linkInertias = worstCaseLinkInertia(config.links, config.masses, payload);
  let peakUtilization = 0;
  let peakJoint: JointName = 'base';
  for (let k = 0; k <= samples; k++) {
    const s = sampleTrajectory(plan, (plan.duration * k) / samples);
    const tauGrav = gravityTorques(s.q, config.links, config.masses, payload);
    JOINT_NAMES.forEach((name, i) => {
      const gb = gearboxes[name];
      const inertia = linkInertias[i] + reflectedInertia(config.motor, gb);
      const required = Math.abs(tauGrav[i]) + inertia * Math.abs(s.qdd[i]);
      const available = availableJointTorque(config.motor, gb, s.qd[i]);
      const util = available > 1e-9 ? required / available : Infinity;
      if (util > peakUtilization) {
        peakUtilization = util;
        peakJoint = name;
      }
    });
  }
  return { peakUtilization, peakJoint, skippedSteps: peakUtilization > 1 };
}

export function computeMetrics(
  config: RobotConfig,
  gearboxes: Record<JointName, GearboxParams>,
  q: JointAngles,
  payload: number,
): TwinMetrics {
  const motor: MotorParams = config.motor;
  const J = jacobian(q, config.links);
  const tauGrav = gravityTorques(q, config.links, config.masses, payload);
  const linkInertias = worstCaseLinkInertia(config.links, config.masses, payload);

  const joints: JointMetrics[] = JOINT_NAMES.map((name, i) => {
    const gb = gearboxes[name];
    const available = availableJointTorque(motor, gb, 0);
    const required = Math.abs(tauGrav[i]);
    const refl = reflectedInertia(motor, gb);
    const totalInertia = refl + linkInertias[i];
    const budget = ACCEL_TORQUE_FRACTION * (available - required);
    return {
      name,
      availableTorque: available,
      requiredTorque: required,
      utilization: available > 1e-9 ? required / available : Infinity,
      maxSpeed: maxJointSpeed(motor, gb),
      maxAccel: Math.max(0, budget / totalInertia),
      reflectedInertia: refl,
      linkInertia: linkInertias[i],
      resolution: jointResolution(motor, gb),
      holdFails: required > available,
    };
  });

  const backlash = JOINT_NAMES.map((n) => gearboxes[n].backlash) as [number, number, number];
  const vmax = joints.map((j) => j.maxSpeed) as [number, number, number];
  const amax = joints.map((j) => j.maxAccel) as [number, number, number];

  return {
    joints,
    backlashErrorTcp: backlashTcpError(J, backlash),
    maxTcpSpeed: maxTcpSpeed(J, vmax),
    singularity: singularityMeasure(q, config.links),
    vmax,
    amax,
  };
}

File: src/simulation/metrics.cpp
#include "metrics.hpp"

#include <algorithm>
#include <cmath>
#include <limits>

#include "../drivetrain/drivetrain.hpp"
#include "../dynamics/dynamics.hpp"
#include "../kinematics/kinematics.hpp"

namespace rt {

namespace {
/** Fraction of the static torque budget reserved for acceleration. */
constexpr double kAccelTorqueFraction = 0.7;
} // namespace

TwinMetrics computeMetrics(const RobotConfig& config,
                           const std::array<GearboxParams, kNumJoints>& gearboxes,
                           const JointAngles& q, double payload) {
  const MotorParams& motor = config.motor;
  const Jacobian J = jacobian(q, config.links);
  const auto tauGrav = gravityTorques(q, config.links, config.masses, payload);
  const auto linkInertias = worstCaseLinkInertia(config.links, config.masses, payload);

  TwinMetrics out{};
  std::array<double, kNumJoints> backlash{};
  for (int i = 0; i < kNumJoints; ++i) {
    const GearboxParams& gb = gearboxes[i];
    const double available = availableJointTorque(motor, gb, 0.0);
    const double required = std::abs(tauGrav[i]);
    const double refl = reflectedInertia(motor, gb);
    const double totalInertia = refl + linkInertias[i];
    const double budget = kAccelTorqueFraction * (available - required);
    out.joints[i] = JointMetrics{
        static_cast<Joint>(i),
        available,
        required,
        available > 1e-9 ? required / available : std::numeric_limits<double>::infinity(),
        maxJointSpeed(motor, gb),
        std::max(0.0, budget / totalInertia),
        refl,
        linkInertias[i],
        jointResolution(motor, gb),
        required > available,
    };
    backlash[i] = gb.backlash;
    out.vmax[i] = out.joints[i].maxSpeed;
    out.amax[i] = out.joints[i].maxAccel;
  }

  out.backlashErrorTcp = backlashTcpError(J, backlash);
  out.maxTcpSpeed = maxTcpSpeed(J, out.vmax);
  out.singularity = singularityMeasure(q, config.links);
  return out;
}

TrajectoryAudit auditTrajectory(const RobotConfig& config,
                                const std::array<GearboxParams, kNumJoints>& gearboxes,
                                const TrajectoryPlan& plan, double payload,
                                int samples) {
  const auto linkInertias = worstCaseLinkInertia(config.links, config.masses, payload);
  double peakUtilization = 0.0;
  Joint peakJoint = Joint::Base;
  for (int k = 0; k <= samples; ++k) {
    const TrajectorySample s = sampleTrajectory(plan, plan.duration * k / samples);
    const auto tauGrav = gravityTorques(s.q, config.links, config.masses, payload);
    for (int i = 0; i < kNumJoints; ++i) {
      const GearboxParams& gb = gearboxes[i];
      const double inertia = linkInertias[i] + reflectedInertia(config.motor, gb);
      const double required = std::abs(tauGrav[i]) + inertia * std::abs(s.qdd[i]);
      const double available = availableJointTorque(config.motor, gb, s.qd[i]);
      const double util = available > 1e-9 ? required / available
                                           : std::numeric_limits<double>::infinity();
      if (util > peakUtilization) {
        peakUtilization = util;
        peakJoint = static_cast<Joint>(i);
      }
    }
  }
  return {peakUtilization, peakJoint, peakUtilization > 1.0};
}

} // namespace rt
